File: app/src/main/java/com/dida/procop/FaqAdapter.java
package com.dida.procop;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;
import java.util.ArrayList;

public class FaqAdapter extends RecyclerView.Adapter<FaqAdapter.MyHolder> {

    ArrayList<FaqData> faqDataList;
    Context context;

    public FaqAdapter(ArrayList<FaqData> _faqDataList, Context _context) {
        this.faqDataList = _faqDataList;
        this.context = _context;
    }

    @NonNull
    @Override
    public FaqAdapter.MyHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.layout_faq_item, parent, false);
        return new FaqAdapter.MyHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull FaqAdapter.MyHolder holder, int position) {
        FaqData faqData = faqDataList.get(position);
        holder.faqTitle.setText(faqData.getTitle());
        holder.faqDescription.setText(faqData.getDescription());

        boolean isExpanded = faqDataList.get(position).isExpanded();
        holder.expandedLayoutFaq.setVisibility(isExpanded ? View.VISIBLE : View.GONE);
        holder.plusMinusImg.setImageResource(isExpanded ? R.drawable.ic_minus : R.drawable.ic_add);
    }

    @Override
    public int getItemCount() {
        return faqDataList.size();
    }

    public class MyHolder extends RecyclerView.ViewHolder{
        TextView faqTitle;
        TextView faqDescription;
        ConstraintLayout expandedLayoutFaq;
        ImageView plusMinusImg;
        public MyHolder(@NonNull View itemView) {
            super(itemView);
            faqTitle = itemView.findViewById(R.id.faqTitle);
            faqDescription = itemView.findViewById(R.id.faqDescription);
            expandedLayoutFaq = itemView.findViewById(R.id.expandedLayoutFaq);
            plusMinusImg = itemView.findViewById(R.id.plusMinusImg);

            faqTitle.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    FaqData faqData = faqDataList.get(getAdapterPosition());
                    faqData.setExpanded(!faqData.isExpanded());
                    notifyItemChanged(getAdapterPosition());
                }
            });
        }
    }
}

File: app/src/main/java/com/androidstarter/ui/home/adapter/FaqAdapter.kt
package com.androidstarter.ui.home.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.androidstarter.R
import com.androidstarter.databinding.LayoutFaqItemBinding
import com.androidstarter.ui.fq.FaqData
import javax.inject.Inject

class FaqAdapter @Inject constructor() :
    RecyclerView.Adapter<FaqAdapter.FaqViewHolder>() {
    var onItemClickListener: ((view: View, position: Int, data: FaqData?) -> Unit)? =
        null
    private var list: MutableList<FaqData> = mutableListOf()
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FaqViewHolder {
        return FaqViewHolder(
            LayoutFaqItemBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(holder: FaqViewHolder, position: Int) {
        holder.bind(list[position])
    }

    override fun getItemCount(): Int {
        return list.size
    }

    fun setList(list: List<FaqData>) {
        this.list.clear()
        this.list.addAll(list)
        notifyDataSetChanged()
    }

    inner class FaqViewHolder(private val itemBinding: LayoutFaqItemBinding) :
        RecyclerView.ViewHolder(itemBinding.root) {

        fun bind(item: FaqData) {
            itemView.setOnClickListener {
                onItemClickListener?.invoke(it, adapterPosition, item)
            }
            itemBinding.faqTitle.text = item.title
            itemBinding.faqDescription.text = item.description

            itemBinding.expandedLayoutFaq.visibility =
                if (item.expanded) View.VISIBLE else View.GONE
            itemBinding.plusMinusImg.setImageResource(if (item.expanded) R.drawable.ic_minus else R.drawable.ic_add)

            itemBinding.faqTitle.setOnClickListener {
                item.expanded = !item.expanded
                notifyItemChanged(adapterPosition)
            }
        }
    }
}